resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java

```diff
@@ -246,8 +246,11 @@ private void flushQueue()
       {
          synchronized (this) {
             if (lastAsyncOperation != null) {
-               // Do not reset the lastAsyncOperation unless the current one is complete
-               if (!lastAsyncOperation.future.isDone()) {
+               lastAsyncOperation.future.complete(null);
```

Comment:
How are we guaranteed that the last bytes have been written at this point?  I thought the point of the `if` check below was to keep things in a state that the last operation would complete successfully.

Reply:
Again the same [comment](https://github.com/resteasy/resteasy/pull/3041#discussion_r826610248) will apply here too. Please find this implementation in the 4.7.3.Final tag [here](https://github.com/resteasy/resteasy/blob/4.7.3.Final/resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java#L236).

Reply:
Here my understanding on this, the lastAsyncOperation is assigned from the queue [here](https://github.com/resteasy/resteasy/blob/ea829f87d61ef60aa66b207c471cbea95b70278e/resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java#L265-L267) and work is called on the lastAsyncOperation(when the outputstream is ready). The work always writes the data to the outputstream. But it may not complete the future due to [this condition](https://github.com/resteasy/resteasy/blob/ea829f87d61ef60aa66b207c471cbea95b70278e/resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java#L71-L73). When that happens,  the OutputStream readiness will be notified via this callback [onWritePossible](https://github.com/resteasy/resteasy/blob/ea829f87d61ef60aa66b207c471cbea95b70278e/resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java#L272-L275) which will call this flushQueue method. So this checks if the lastAsyncOperation is not null and if it is not, then it completes the future.

Reply:
[This](https://tomcat.apache.org/tomcat-9.0-doc/api/org/apache/catalina/connector/CoyoteOutputStream.html#isReady()) suggests that a call `onWritePossible` implicitly guarantees that we can safely mark the task as done.

Reply:
Thanks for the link.  That really clears up a lot.